### PR TITLE
Improve pattern recognition of Concat chains

### PR DESCRIPTION
**Pull request by jamesqo**
**TL;DR** Without adding any new types or fields, we now recognize patterns of the form `a.Concat(b.Concat(c))` (where we hold a reference to the first enumerable to be concatenated) in addition to `a.Concat(b).Concat(c)` (where we hold a reference to the last one). We can also optimize for mixing of the two, e.g. `a.Concat(b.Concat(c)).Concat(d)`.

### Summary

- ConcatIterator has 2 new abstract methods: `ConcatAfter` and `ConcatBefore`. `ConcatAfter` is just a renaming what was previously `Concat` (produces an iterator with the enumerable coming after the rest of the nodes). `ConcatBefore` produces an iterator with the enumerable coming before the rest of the nodes.

- Each iterator's implementation of ConcatBefore is basically exactly the same as ConcatAfter, with one small exception: we set the sign bit of `_index` to mark it as a 'prepended' node, or that the enumerable it holds comes before the other nodes.

  - Implementing `GetEnumerable` is now slightly tricky, since `_index & int.MaxValue` no longer represents the index of the enumerable, but rather the order in which it was added. For example, consider this code:

    ```cs
    List<T> Single<T>(T item) => new List<T> { item };

    Single(3).Concat(Single(2).Concat(Single(0).Concat(Single(1))))).Concat(Single(4))
    ```

    This will produce a node representation with these indices (x' stands for `x | int.MinValue` and represents a prepended node):

    `3' 2' [0 1] 4`

    Where 4 references 3', 3' references 2', and 2 references \[0 1\] (the first 2 collections). As you can see the lower 31 bits of `_index` for 3' and 2' will be 3 and 2, but they are actually indices 0 and 1 from the viewpoint of `GetEnumerable`.

    - To solve this problem, we say that `GetEnumerable(n)` of an appended node should be the items held by that node if `n == _index`, otherwise it should be `_linked.GetEnumerable(n)`. For prepended nodes, it should be the items held by that node if `n == 0`, otherwise it should be `_linked.GetEnumerable(n - 1)`. (This is all implemented iteratively to avoid SOs of course.)

      - In the above example, if we wanted `GetEnumerable(1)`, we would first check 4 == 1. Then we would hop to 3', and check 1 == 0, then decrement 1. Then we would hop to 2', find 0 == 0 and return the items at 2'.

    - Another problem we face is that the `index > _index` check no longer works if the head of the linked list is prepended-- we have to find the first appended node (which will hold the last concatee, and therefore the greatest index) before making that check.

- Other changes

  - `Concatenate.cs` has been renamed to `Concat.cs`. It was previously named `Concatenate` b/c it also housed Append/Prepend, but those have since been moved to another file.

  - Fixed existing bug where we ignore the `arrayIndex` in `Concat2CollectionIterator.CopyTo` and just use 0. (Note that this is dormant currently because w/ existing builds we only pass in 0 for `arrayIndex`.)

  - `_previous` / `_next` / `_nextIndex` are no longer accurate field names, so I changed them to `_linked` / `_concatee` / `_index`.

  - `ConcatNCollectionIterator.CopyBefore` has been eliminated-- we can't do `CopyTo(T[], int)` efficiently since we have to walk all the way to the first prepended node, and with `CopyBefore(T[], int)` we likewise have to find the first appended node. My idea was to have it accept a `startIndex` and a `count`; items from prepended nodes will be copied starting at `startIndex`, and ones from appended nodes towards the end.

  - Added strenuous testing for the new changes, which helped me uncover the bug I mentioned earlier.

@VSadov @stephentoub @justinvp PTAL

### CODE REVIEW

**Comment by jamesqo** [comment]
I didn't post it in the description b/c it was getting too long, but here is some performance data:

- [Regression](https://gist.github.com/jamesqo/68a5cf465da728ef51a7089285e57a51) (both sides are enumerables suffer ~10% due to an extra typecast)

- ["Worst" best case](https://gist.github.com/jamesqo/9606804646689cb6e9a1ec8e1bc28da3) (all inputs are ICollections, but only length-1 so CopyTo isn't that helpful)

**Comment by stephentoub** [comment]
> both sides are enumerables suffer ~10% due to an extra typecast

This is _the_ most common case. Further harming it for a much less common case is not a good tradeoff.

**Comment by jamesqo** [comment]
@stephentoub I made a new benchmark in which we actually iterate through the (empty) Concat iterator-- https://gist.github.com/jamesqo/24948e533e3f01eea6a1d236cb920b09

There is basically no difference between the old/new timings, both are around ~3.2s.

**Comment by stephentoub** [comment]
> Any reason that a.Concat(b.Concat(c)) should be less common than a.Concat(b).Concat(c)?

Because it's what people write?  From the codebases I've explored looking for usage, by far the most common usage is simply combining two enumerables of data, e.g. ```a.Concat(b)```.  After that, the most common is combining multiple sets, and it's much more natural to write that as ```a.Concat(b).Concat(c).Concat(d)``` rather than as ```a.Concat(b.Concat(c.Concat(d)))```, so that's what people do.  I'm sure there are cases where it's done, I've just not seen it to be common.

Prior to very recently, Concat was a super simple implementation:
https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,692
```C#
static IEnumerable<TSource> ConcatIterator<TSource>(IEnumerable<TSource> first, IEnumerable<TSource> second) {
    foreach (TSource element in first) yield return element;
    foreach (TSource element in second) yield return element;
}
```
We got some complaints that when chains of concats were being used (even short chains but with large numbers of elements), the time spent could increase significantly, so we addressed those concerns by changing the algorithm for this case from being O(N^2) to O(N).  That was worthwhile because it addressed a real-world performance issue that was affecting real code in real applications to a measurable extent.

Each of the changes to concat since has, IMHO, had much less significant ROI.  And while I appreciate the obvious effort you're putting into these changes, any such change like this has risk, adds complexity and cost, and requires effort not just from you but from everyone else who needs to review and think through the changes before they're accepted into the codebase, and then any subsequent effort involved in maintaining.

That's why you're seeing me being skeptical of such a change.  It's good work, but from what I've seen, it addresses a relatively rare coding pattern, with a fix that adds cost (even if small) to common coding patterns, that adds several hundred lines of code, etc.  I would personally find it more valuable if you would focus your energies on areas that are known to have a real impact on real codebases, rather than improving the throughput of a microbenchmark for some particular code pattern that may or may not affect real apps in a meaningful way.

If you can point to some examples on GitHub or in referencesource.microsoft.com where this change does make a measurable perf impact, I'd be happy to take a look and would be more interested.

**Comment by jamesqo** [comment]
@stephentoub Thank you for your response.

> If you can point to some examples on GitHub or in referencesource.microsoft.com where this change does make a measurable perf impact, I'd be happy to take a look and would be more interested.

Since the xUnit test runner runs on the same locally-built assemblies that are being tested, I decided to test how frequently `ConcatBefore` is called by the test runner. I set up [this code](https://github.com/jamesqo/corefx/commit/eca388294867faa42dc25e17e2ee499b6ebde1d0) to break the debugger every time the number of usages passed a power of 2, and did `./corerun -d xunit.console.netcore.exe ...` and attached to the process in VS.

It looks like [this line](https://github.com/xunit/xunit/blob/master/src/xunit.execution/Sdk/Reflection/ReflectionAttributeInfo.cs#L102) is calling `ConcatBefore` somewhere between 2048 and 4096 times before the tests are actually run. Proof: [image 1](http://imgur.com/a/0Oujv), [image 2](http://imgur.com/a/AIpWV). The second image shows that `ConcatNCollectionIterator.ConcatBefore` is being called over 1000 times.

> After that, the most common is combining multiple sets, and it's much more natural to write that as `a.Concat(b).Concat(c).Concat(d)` rather than as `a.Concat(b.Concat(c.Concat(d)))`, so that's what people do. I'm sure there are cases where it's done, I've just not seen it to be common.

It's worth noting that there are other cases in which the second syntax is more 'natural' than the first-- for example

```cs
var items = b.Concat(c);

if (flag)
{
    items = a.Concat(items);
}

// vs

var items = flag ? a.Concat(b).Concat(c) : b.Concat(c);
```

**Comment by jamesqo** [comment]
On second thought, I've changed my mind on this PR. @stephentoub as you mention

> I would personally find it more valuable if you would focus your energies on areas that are known to have a real impact on real codebases, rather than improving the throughput of a microbenchmark for some particular code pattern that may or may not affect real apps in a meaningful way.

I had initially started work on this PR under the assumption that `a.Concat(b.Concat(c))` was about equally as common as `a.Concat(b).Concat(c)`. However, as you mention, the second form is more common by a large margin; at a first glance, there are numerous examples on referencesource.microsoft.com where the second pattern is used, and none of the first one.

With that in mind, I'm closing this PR. Was still very fun to work on though :smile:

**Comment by stephentoub** [comment]
Thanks!

**Comment by jamesqo** [comment]
@stephentoub I'm curious, how does one collect telemetry on these kinds of things? For example, is it possible to perform a build of Roslyn with custom corefx binaries and see how many times a method is being called / how much time it's taking up? Thanks.

**Comment by stephentoub** [comment]
> For example, is it possible to perform a build of Roslyn with custom corefx binaries and see how many times a method is being called / how much time it's taking up?

Sure.  You just need to overwrite the System.Linq.dll that's used with one you build.  Should work fine.  As for measuring, you could use a profiler like PerfView, you could add your own instrumentation into the binary you deploy, etc.
